packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -459,6 +477,29 @@ LightningElement.prototype = {
         return vm.shadowRoot;
     },
 
+    get refs(): TemplateRefs {
+        const vm = getAssociatedVM(this);
+
+        if (isNull(vm.refs)) {
+            return EMPTY_REFS;
+        }
+
+        // The Template refs can be cached based on the VM refs, since the VM refs
+        // are recreated every time the template is rendered.
+        let refs = refsCache.get(vm.refs);
```

Comment:
the note from above will make this a lot more readable... `let refs = refsCache.get(vm.refNodes);`